test/validation/api/ipsec/ipsec.c

```diff
@@ -815,26 +947,46 @@ static int ipsec_send_out_one(const ipsec_test_part *part,
 			ev = recv_event(queue, 0);
 			if (ODP_EVENT_INVALID != ev) {
 				CU_ASSERT(odp_event_is_valid(ev) == 1);
+
 				CU_ASSERT_EQUAL(ODP_EVENT_PACKET,
 						odp_event_types(ev, &subtype));
-				CU_ASSERT_EQUAL(ODP_EVENT_PACKET_BASIC,
-						subtype);
-				CU_ASSERT(!part->out[i].status.error.all);
+
+				if (ODP_EVENT_PACKET_IPSEC != subtype) {
+					CU_ASSERT_EQUAL(ODP_EVENT_PACKET_BASIC,
+							subtype);
+					CU_ASSERT(!part->out[i].status.error.all);
+				}
 
 				pkto[i] = odp_packet_from_event(ev);
 				CU_ASSERT_FATAL(pkto[i] != ODP_PACKET_INVALID);
+
+				if (part->out[i].sa_expiry != IPSEC_TEST_EXPIRY_IGNORED)
+					ipsec_status_event_get(sa, part->out[i].sa_expiry);
+
 				i++;
 				continue;
 			}
 
 			ev = recv_event(suite_context.queue, 0);
 			if (ODP_EVENT_INVALID != ev) {
+				odp_event_type_t ev_type;
+
 				CU_ASSERT(odp_event_is_valid(ev) == 1);
+				ev_type = odp_event_types(ev, &subtype);
+
+				if ((ODP_EVENT_IPSEC_STATUS == ev_type) &&
+				    part->out[i].sa_expiry != IPSEC_TEST_EXPIRY_IGNORED) {
+					ipsec_status_event_handle(ev, sa, part->out[i].sa_expiry);
+					continue;
+				}
+
 				CU_ASSERT_EQUAL(ODP_EVENT_PACKET,
-						odp_event_types(ev, &subtype));
+						ev_type);
 				CU_ASSERT_EQUAL(ODP_EVENT_PACKET_IPSEC,
 						subtype);
-				CU_ASSERT(part->out[i].status.error.all);
+				if (part->out[i].sa_expiry != IPSEC_TEST_EXPIRY_HARD_PKT &&
+				    part->out[i].sa_expiry != IPSEC_TEST_EXPIRY_HARD_BYTE)
+					CU_ASSERT(part->out[i].status.error.all);
```

Comment:
This can be a bit hard to understand. It is related to the somewhat unclear maning of the sa_expiry field. Maybe it would be clearer if the expected status is checked to contain an error always except when sa_expiry is IGNORE, in which case both hard expiry and success are ok (with a bit tweaked meaning of IGNORE and with the expiry test setting the expected error flag when it in fact expects an error.

Reply:
Agreed. I think in this function (ipsec_send_out_one()) if the SA expiry is NONE, we can assert on status.error.all.
If sa_expiry is set to any value other than NONE, the result status structure is subsequently getting checked in ipsec_check_out() via ipsec_check_sa_expiry().

I had introduced the new function ipsec_check_sa_expiry() to check the status fields for better code indentation.